File: libs/language/src/lib/provider/SequentialNodeProvider.ts
import { NodeProvider } from '../NodeProvider';
import { BlueNode } from '../model';

/**
 * A NodeProvider that tries to fetch nodes sequentially from a list of providers
 * Returns the result from the first provider that returns a non-empty result
 */
export class SequentialNodeProvider extends NodeProvider {
  private nodeProviders: NodeProvider[];

  constructor(nodeProviders: NodeProvider[]) {
    super();
    this.nodeProviders = nodeProviders;
  }

  override fetchByBlueId(blueId: string): BlueNode[] {
    for (const provider of this.nodeProviders) {
      const nodes = provider.fetchByBlueId(blueId);
      if (nodes && nodes.length > 0) {
        return nodes;
      }
    }
    return [];
  }

  // Override fetchFirstByBlueId for more efficient implementation
  // In Java, this would call the default implementation, but we optimize here
  override fetchFirstByBlueId(blueId: string): BlueNode | null {
    for (const provider of this.nodeProviders) {
      const node = provider.fetchFirstByBlueId(blueId);
      if (node) {
        return node;
      }
    }
    return null;
  }

  getNodeProviders(): NodeProvider[] {
    return this.nodeProviders;
  }
}

File: references/java-sdk/src/main/java/blue/language/provider/SequentialNodeProvider.java
package blue.language.provider;

import blue.language.model.Node;
import blue.language.NodeProvider;

import java.util.Arrays;
import java.util.List;
import java.util.Objects;

public class SequentialNodeProvider implements NodeProvider {
    private List<NodeProvider> nodeProviders;

    public SequentialNodeProvider(List<NodeProvider> nodeProviders) {
        this.nodeProviders = nodeProviders;
    }

    public SequentialNodeProvider(NodeProvider... nodeProviders) {
        this.nodeProviders = Arrays.asList(nodeProviders);
    }

    @Override
    public List<Node> fetchByBlueId(String blueId) {
        return nodeProviders.stream()
                .map(provider -> provider.fetchByBlueId(blueId))
                .filter(Objects::nonNull)
                .findFirst()
                .orElse(null);
    }

    public List<NodeProvider> getNodeProviders() {
        return nodeProviders;
    }
}